src/backend/distributed/executor/merge_executor.c

```diff
@@ -0,0 +1,422 @@
+	 * on shards with connections.
+	 */
+	Task *task = NULL;
+	foreach_ptr(task, taskList)
+	{
+		uint64 shardId = task->anchorShardId;
+		bool shardModified = false;
+
+		hash_search(shardStateHash, &shardId, HASH_FIND, &shardModified);
+		if (shardModified)
+		{
+			prunedTaskList = lappend(prunedTaskList, task);
+		}
+	}
+
+	if (prunedTaskList == NIL)
+	{
+		/* No task to execute */
+		return;
+	}
+
+	ereport(DEBUG1, (errmsg("Execute MERGE task list")));
+	bool randomAccess = true;
+	bool interTransactions = false;
+	Assert(scanState->tuplestorestate == NULL);
+	scanState->tuplestorestate = tuplestore_begin_heap(randomAccess, interTransactions,
+													   work_mem);
+	TupleDesc tupleDescriptor = ScanStateGetTupleDescriptor(scanState);
+	TupleDestination *tupleDest =
+		CreateTupleStoreTupleDest(scanState->tuplestorestate, tupleDescriptor);
+	uint64 rowsMerged =
+		ExecuteTaskListIntoTupleDest(ROW_MODIFY_COMMUTATIVE, prunedTaskList,
+									 tupleDest, hasReturning);
+	executorState->es_processed = rowsMerged;
+}
+
+
+/*
+ * sourceResultPartitionColumnIndex collects all Join conditions from the
+ * ON clause and verifies if there is a join, either left or right, with
+ * the distribution column of the given target. Once a match is found, it
+ * returns the index of that match in the source's target list.
+ */
+static int
+sourceResultPartitionColumnIndex(Query *mergeQuery, List *sourceTargetList,
+								 CitusTableCacheEntry *targetRelation)
+{
+	/* Get all the equal(=) expressions from the ON clause */
+	List *mergeJoinConditionList = QualifierList(mergeQuery->jointree);
+
+	/*
+	 * Now that we have gathered all the _equal_ expressions in the list, we proceed
+	 * to match the one that joins with the distribution column of the target.
+	 */
+	Var *targetColumn = targetRelation->partitionColumn;
+	List *sourceJoinColumns = NIL;
+	Node *joinClause = NULL;
+	foreach_ptr(joinClause, mergeJoinConditionList)
+	{
+		Node *leftOperand;
+		Node *rightOperand;
+
+		if (!BinaryOpExpression((Expr *) joinClause, &leftOperand, &rightOperand))
+		{
+			continue;
+		}
+
+		if (equal(leftOperand, targetColumn) && IsA(rightOperand, Var))
+		{
+			sourceJoinColumns = lappend(sourceJoinColumns, rightOperand);
+		}
+		else if (equal(rightOperand, targetColumn) && IsA(leftOperand, Var))
+		{
+			sourceJoinColumns = lappend(sourceJoinColumns, leftOperand);
+		}
+	}
+
+	int partitionColumnIndex = -1;
+	if (sourceJoinColumns)
+	{
+		/* Use the Var that's being used in INSERT clause */
+		Var *sourceJoinColumn =
+			ErrorIfInsertNotMatchTargetDistributionColumn(
+				targetRelation->relationId, mergeQuery,
+				sourceJoinColumns);
+
+		/* If no INSERT action is present, we can use one of the Var */
+		if (!sourceJoinColumn)
+		{
+			sourceJoinColumn = linitial(sourceJoinColumns);
+		}
+
+		/*
+		 * Find the index in the source list that will be utilized
+		 * for colocation-redistribution.
+		 */
+		partitionColumnIndex =
```

Comment:
minor: there is some naming confusion about partitionColumnIndex/DistributionColumnIndex.

I wonder if we should not use any wording that might feel like we are dealing with any distributed table partition/dist key. Instead, we are dealing with "source RTE repartition column" or something along the line.

So, not sure, but probably better to avoid "distribution/partition key" terminology here. What is your take? maybe `sourceResultRepartitionColumnIndex` (and variable names along that line)

Reply:
Yeah I noticed that mix-up, and changed it to `distributionColumnIndex` (didn't push it yet). But I guess that's still misleading. I've put it as `sourceResultRepartitionColumnIndex`